### PR TITLE
David sort component

### PR DESCRIPTION
**Pull request by da411d**
# SortList component
SortList component sorts list of tasks by name and ptiority
## Usage
Put the selection widget in some place:
```
<SortList
  storageKey="someKey"
  sortIconColor="#abc123"
/>
```

**storageKey** - the key for LocalStorage, where we will save the state of sort field and direction.
It should be unique exactly for every list.
**sortIconColor** - color of sort icon

Then, import `sort` function from `utils.js`
```
const sortedTasks = sort(tasksForSorting, field, direction);
```

### CODE REVIEW

File: src/components/Dashboard/SortList/SortList.jsx [hunk 1]
```diff
@@ -27,9 +27,48 @@ export default class SortList extends Component {
   };
 
   render() {
-    const { children, sortIconColor } = this.props;
+    const { children, sortIconColor, fields } = this.props;
     const { currentField, currentDirection } = this.state;
 
+    const sortVariantsList = fields.map((field) => {
```
Comment (mishalisovskiy) [inline src/components/Dashboard/SortList/SortList.jsx]: No need to map four buttons!!!

**Comment by mishalisovskiy** [comment]
The previous pull request shows the history of all the problems with this component. Please push new commits to that PR

**Comment by da411d** [comment]
New version with fixes and improvements is in progress and will be done in several days.

File: src/components/Dashboard/SortList/SortList.jsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import * as sortListActions from '../../../actions/sortListActions';
+import './SortList.scss';
+import { ReactComponent as SortIcon } from './sort.svg';
+import { ReactComponent as SortIconAsc } from './sort-asc.svg';
+import { ReactComponent as SortIconDesc } from './sort-desc.svg';
+
+class SortList extends Component {
+  state = {};
```
Comment (mishalisovskiy) [inline src/components/Dashboard/SortList/SortList.jsx]: Why the state should be outside the costructor?

File: src/components/Dashboard/SortList/SortList.scss [hunk 3]
```diff
@@ -0,0 +1,43 @@
+.btn{
```
Comment (mishalisovskiy) [inline src/components/Dashboard/SortList/SortList.scss]: I think it was made clear several times, we use ident with 2 spaces, not four.

File: src/components/Dashboard/SortList/sortUtils.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    let data = localStorage.getItem('SortList_Data') || '{}';
+    data = JSON.parse(data);
+    return data[key];
+  },
+
+  set: (key, value) => {
+    let data = localStorage.getItem('SortList_Data') || '{}';
+    data = JSON.parse(data);
+    data[key] = value;
+    data = JSON.stringify(data);
```
Comment (mishalisovskiy) [inline src/components/Dashboard/SortList/sortUtils.js]: Using the same variable for the entire block and mutating it in the process steers far away from recommended practices.

File: src/reducers/sortListReducer.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { SET_SORT } from '../actions/actionTypes';
+
+export default (state = {}, action) => {
+  const {
+    key,
+    field,
+    direction,
+  } = action;
```
Comment (mishalisovskiy) [inline src/reducers/sortListReducer.js]: Read the documentation to see what keys are allowed in a reducer.

File: src/reducers/sortListReducer.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  switch (action.type) {
+    case SET_SORT:
+      return {
+        ...state,
```
Comment (mishalisovskiy) [inline src/reducers/sortListReducer.js]: Do you really want to keep the previous state? All the sorting component does is changes the sorting type. So, your state should change completely.

File: src/components/Dashboard/SortList/SortList.jsx (hunk 2 above)
Comment (da411d) [reply src/components/Dashboard/SortList/SortList.jsx]: oops. This variable is not needed. I forgot to remove it.

File: src/reducers/sortListReducer.js (hunk 6 above)
Comment (da411d) [reply src/reducers/sortListReducer.js]: Yes, i really want to keep the previous state.
The component saves fort direction for all columns:
```
sortListReducer: {
    'DashboardId:Column': {
      field: 'name',
      direction: 'DESC'
    },
    'DashboardId:Column': {
      field: 'priority',
      direction: 'ASC'
    },
    'DashboardId:Column': {
      field: 'name',
      direction: 'ASC'
    },
}
```
So, if we change the state of one of columns, we should keep state of others.

File: src/components/Dashboard/SortList/SortList.scss (hunk 3 above)
Comment (da411d) [reply src/components/Dashboard/SortList/SortList.scss]: Sorry i didn't notice.
Usually, linter highlights problems with ident, so I did not pay attention.

File: src/components/Dashboard/TasksColumn/TasksColumn.jsx [hunk 7]
```diff
@@ -26,14 +26,14 @@ class TasksColumn extends Component {
...
     } = this.props;
 
     const sortIconColor = loaderColor(title);
 
     const storageKey = `${document.URL.split('/').pop()}:${title}`;
     const sortState = allSortData[storageKey] || {};
-    let sortedTasks = tasks;
+    let sortedTasks = filteredTasks;
     if (sortState.direction !== 'NONE') {
       sortedTasks = sort(sortedTasks, sortState.field, sortState.direction);
```
Comment (mishalisovskiy) [inline src/components/Dashboard/TasksColumn/TasksColumn.jsx]: Mutating a variable and throwing it into a function as an argument isn't the best idea. Please rewrite without `let` and without this -- `sortedTasks = sort(sortedTasks`

File: src/actions/sortListActions.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    direction,
+  });
+};
+
+const loadSort = key => dispatch => {
+  const data = storage.get(key) || {};
+  const { field, direction } = data;
+
+  dispatch({
+    type: SET_SORT,
```
Comment (mishalisovskiy) [inline src/actions/sortListActions.js]: I can see only one type in two functions, which makes one of those redundant. Use dynamic dispatching techniques and refactor duplicate code

File: src/components/Dashboard/SortList/SortList.jsx [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                variant=""
+                className="rounded-0 text-left disabled btn-no-outline"
+              >
+                Name
+              </Button>
+              <Button
+                variant="light"
+                active={field === 'name' && direction === 'ASC'}
+                className="rounded-0 btn-no-outline flex-grow-0"
+                onClick={() => this.onClickCallback('name', 'ASC')}
```
Comment (mishalisovskiy) [inline src/components/Dashboard/SortList/SortList.jsx]: This function name tells me nothing of how it works. Please change

File: src/components/Dashboard/SortList/SortList.jsx [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+  constructor(props) {
+    super(props);
+    const { storageKey, sortListActions: { loadSort } } = this.props;
+
+    loadSort(storageKey);
+  }
+
+  onClickCallback = (newField, newDirection) => {
+    const { storageKey, allSortData, sortListActions: { setSort } } = this.props;
```
Comment (mishalisovskiy) [inline src/components/Dashboard/SortList/SortList.jsx]: `allSortData` -- what kind of data is this? The name is too vague for an outside reviewer to understand.

File: src/components/Dashboard/TasksColumn/TasksColumn.jsx (hunk 7 above)
Comment (da411d) [reply src/components/Dashboard/TasksColumn/TasksColumn.jsx]: As you see in [sortUtils.js:2](https://github.com/Lv398NodeJS/add-a-thing/blob/67b977596c9106d35f0f370432d23310055bac93/src/components/Dashboard/SortList/sortUtils.js#L2), we are not mutating array.

Comment (mishalisovskiy) [reply src/components/Dashboard/TasksColumn/TasksColumn.jsx]: Change it to `const` and see if it works

File: src/components/Dashboard/SortList/SortList.jsx (hunk 10 above)
Comment (da411d) [reply src/components/Dashboard/SortList/SortList.jsx]: allSortData - it stores sort settings for all dashboards and all columns.

Comment (mishalisovskiy) [reply src/components/Dashboard/SortList/SortList.jsx]: I don't want you to explain it. Your variable names should be clear enough without any explanation.

File: src/components/Dashboard/SortList/SortList.scss [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        }
+    }
+}
+
+img.icon{
+    height: 1.5em;
+}
+
+
+.dropdown-pull-up{
```
Comment (mishalisovskiy) [inline src/components/Dashboard/SortList/SortList.scss]: It is customary to put spaces after selectors in CSS. This code is a little ugly.

File: src/components/Dashboard/SortList/SortList.scss [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+            background-color: #d8d9da !important;
+            border-color: #d8d9da !important;
+        }
+    }
+}
+
+.dropdown-toggle{
+    &.no-arrow{
+        &::after{
+            display: none !important;
```
Comment (mishalisovskiy) [inline src/components/Dashboard/SortList/SortList.scss]: Usually, we can do away without `!important`. Is there no other way?

File: src/actions/sortListActions.js [hunk 13]
```diff
@@ -15,16 +15,10 @@ const setSort = (key, field, direction) => dispatch => {
   });
 };
 
-const loadSort = key => dispatch => {
+const loadSort = key => (dispatch) => {
```
Comment (mishalisovskiy) [inline src/actions/sortListActions.js]: No need for this very function -- `loadSort`

File: src/components/Dashboard/SortList/SortList.jsx (hunk 10 above)
Comment (da411d) [reply src/components/Dashboard/SortList/SortList.jsx]: I don't explain. I just show that variable name shows the sense of its content.

File: src/components/Dashboard/SortList/sortUtils.js [hunk 14]
```diff
@@ -32,10 +32,9 @@ export const storage = {
   },
 
   set: (key, value) => {
-    let data = localStorage.getItem('SortList_Data') || '{}';
-    data = JSON.parse(data);
-    data[key] = value;
-    data = JSON.stringify(data);
-    localStorage.setItem('SortList_Data', data);
+    const data = localStorage.getItem('SortList_Data') || '{}';
+    const object = JSON.parse(data);
```
Comment (mishalisovskiy) [inline src/components/Dashboard/SortList/sortUtils.js]: What are `data` and `object`. Stop giving vague names to variables, and for once give it at least some thought

File: src/components/Dashboard/SortList/SortList.jsx (hunk 10 above)
Comment (mishalisovskiy) [reply src/components/Dashboard/SortList/SortList.jsx]: It doesn't, David, or I wouldn't be asking. Think how to change the variable so it is self-explanatory

File: src/components/Dashboard/SortList/sortUtils.js (hunk 14 above)
Comment (deviun) [reply src/components/Dashboard/SortList/sortUtils.js]: data -> sortListDataRaw
object -> sortListData

File: src/components/Dashboard/TasksColumn/TasksColumn.jsx [hunk 15]
```diff
@@ -23,7 +25,18 @@ class TasksColumn extends Component {
...
+    const {
+      title, filteredTasks, loading, allSortData = {},
+    } = this.props;
+
+    const sortIconColor = loaderColor(title);
+
+    const storageKey = `${document.URL.split('/').pop()}:${title}`;
+    const sortState = allSortData[storageKey] || {};
+    let sortedTasks = filteredTasks;
+    if (sortState.direction !== 'NONE') {
```
Comment (deviun) [inline src/components/Dashboard/TasksColumn/TasksColumn.jsx]: Unknown string constant, move this to constants file with description

File: src/components/Dashboard/TasksColumn/TasksColumn.jsx (hunk 7 above)
Comment (deviun) [reply src/components/Dashboard/TasksColumn/TasksColumn.jsx]: sortedTasks may sorted inside himself, how works native .sort method.
Screenshot 2019-05-31 at 12 18 50 AM

Comment (deviun) [reply src/components/Dashboard/TasksColumn/TasksColumn.jsx]: Also, u can use mapper function for sort

sortedTasks.sort(yourMapper(field, direction))

const yourMapper = (field, direction) => (a, b) => { // code mapper }

File: src/components/Dashboard/SortList/sortUtils.js (hunk 4 above)
Comment (mishalisovskiy) [reply src/components/Dashboard/SortList/sortUtils.js]: David, don't 'resolve' unresolved issues.

File: src/components/Dashboard/TasksColumn/TasksColumn.jsx (hunk 7 above)
Comment (mishalisovskiy) [reply src/components/Dashboard/TasksColumn/TasksColumn.jsx]: You are updating a variable to a result of a function where this variable is an argument. Does this sound in any way okay to you?

File: src/actions/sortListActions.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+  dispatch({
+    type: SET_SORT,
+    key,
+    field,
+    direction,
+  });
+};
+
+const loadSort = key => dispatch => {
```
Comment (mishalisovskiy) [inline src/actions/sortListActions.js]: This function is unnecessary. Please change the one above and use it for all the Redux communication

File: src/components/Dashboard/SortList/sortUtils.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    compareB = String(['low', 'medium', 'high'].indexOf(compareB));
+  }
+
+  const directionMultiplier = direction === 'ASC' ? 1 : -1;
+  return compareA.localeCompare(compareB) * directionMultiplier;
+};
+
+export const storage = {
+  get: (key) => {
+    let data = localStorage.getItem('SortList_Data') || '{}';
```
Comment (mishalisovskiy) [inline src/components/Dashboard/SortList/sortUtils.js]: This `data` isn't the best option too. I'm sure you can come up with a name that immediately identifies the variable for what it is

Comment (deviun) [reply src/components/Dashboard/SortList/sortUtils.js]: sortListDataRaw = get from local storage
sortListData = parse local storage data

File: src/components/Dashboard/TasksColumn/TasksColumn.jsx [hunk 18]
```diff
@@ -88,7 +88,7 @@ class TasksColumn extends Component {
 const mapStateToProps = ({ mainContainerReducer: { taskListRef, loading }, sortListReducer }) => ({
   taskListRef,
   loading,
-  allSortData: sortListReducer,
+  allSortSettings: sortListReducer,
```
Comment (mishalisovskiy) [inline src/components/Dashboard/TasksColumn/TasksColumn.jsx]: This is awesome :) Thank you.

**Pull request merged**